components/i2cdev/i2cdev_master.c

```diff
@@ -0,0 +1,197 @@
+            {
+                i2c_del_master_bus(states[i].bus);
+            }
+            states[i].installed = false;
+            SEMAPHORE_GIVE(i);
+        }
+#if !CONFIG_I2CDEV_NOLOCK
+        vSemaphoreDelete(states[i].lock);
+#endif
+        states[i].lock = NULL;
+    }
+    return ESP_OK;
+}
+
+esp_err_t _i2c_dev_probe(i2c_dev_t *dev, i2c_dev_type_t operation_type)
+{
+    if (!dev) return ESP_ERR_INVALID_ARG;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    esp_err_t res = i2c_setup_port(dev);
+    if (res == ESP_OK)
+    {
+        res = i2c_master_probe(dev->bus, dev->device_config.device_address, CONFIG_I2CDEV_TIMEOUT);
+    }
+
+    SEMAPHORE_GIVE(dev->port);
+
+    return res;
+}
+
+esp_err_t _i2c_dev_read(i2c_dev_t *dev, const void *out_data, size_t out_size, void *in_data, size_t in_size)
+{
+    if (!dev || !in_data || !in_size) return ESP_ERR_INVALID_ARG;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    esp_err_t res = i2c_setup_port(dev);
+    if (res == ESP_OK)
+    {
+
+        if (out_data && out_size)
+        {
+            res = i2c_master_transmit_receive(dev->device, (void *)out_data, out_size, in_data, in_size, CONFIG_I2CDEV_TIMEOUT);
+        }
+        else
+        {
+            res = i2c_master_receive(dev->device, (void *)out_data, out_size, CONFIG_I2CDEV_TIMEOUT);
+        }
+        if (res != ESP_OK)
+        {
+            ESP_LOGE(TAG, "Could not read from device [0x%02x at %d]: %d (%s)", dev->addr, dev->port, res, esp_err_to_name(res));
+        }
+    }
+
+    SEMAPHORE_GIVE(dev->port);
+    return res;
+}
+
+esp_err_t _i2c_dev_write(i2c_dev_t *dev, const void *out_reg, size_t out_reg_size, const void *out_data, size_t out_size)
+{
+    if (!dev || !out_data || !out_size) return ESP_ERR_INVALID_ARG;
+    uint8_t *write_buffer = NULL;
+    size_t write_buffer_size = 0;
+    esp_err_t res = ESP_FAIL;
+    bool with_register_address = out_reg && out_reg_size;
+
+    SEMAPHORE_TAKE(dev->port);
+
+    write_buffer_size = with_register_address ? out_reg_size + out_size : out_size;
+    write_buffer = malloc(write_buffer_size);
+    if (!write_buffer)
+    {
+        ESP_LOGE(TAG, "_i2c_dev_write: malloc() failed: size %d", write_buffer_size);
+        res = ESP_ERR_NO_MEM;
+        goto fail;
+    }
+    res = i2c_setup_port(dev);
+    if (res != ESP_OK)
+    {
+        ESP_LOGE(TAG, "i2c_setup_port(): %s", esp_err_to_name(res));
+        goto fail;
+    }
+
+    if (with_register_address)
+    {
+        memcpy(write_buffer, out_reg, out_reg_size);
+        memcpy(write_buffer + out_reg_size, out_data, out_size);
+
+    }
+    else
+    {
+        memcpy(write_buffer, out_data, out_size);
+    }
+    res = i2c_master_transmit(dev->device, write_buffer, out_reg_size + out_size, CONFIG_I2CDEV_TIMEOUT);
+    if (res != ESP_OK)
+    {
+        ESP_LOGE(TAG, "Could not write to device [0x%02x at %d]: %d (%s)", dev->addr, dev->port, res, esp_err_to_name(res));
+        goto fail;
+    }
+fail:
+    free(write_buffer);
```

Comment:
Are you only freeing on fail? Doesn't this leak memory when the transaction succeeds?

Reply:
IIRC, there is no other path to return after `malloc()` even if succeeded or failed.

Reply:
Right, forgot that, oops